rfcs/0181-list-index-syntax.md

```diff
@@ -0,0 +1,290 @@
+error: attempt to call something which is not a function but a list
+
+       at «string»:1:1:
+
+            1| someExpr.foo.bar.qux.0
+             | ^
+```
+
+Of course that doesn't work.
+I don't actually type that.
+What I actually do is hit up-arrow, then hit Home, then type `builtins.elemAt`, then hit End, then type ` 0`.
+
+```
+nix-repl> builtins.elemAt someExpr.foo.bar.qux 0
+{ greatMoreStuff = { ... }; }
+```
+
+Now what do I get to do?
+That's right, hit up-arrow, then hit Home, then type `(`, then hit End, then type `).greatMoreStuff`.
+
+---
+
+When writing Nix code, it is relatively uncommon to want to index into a list, and `builtins.elemAt` suffices.
+When exploring data in a REPL, however, indexing into lists is more common, and the above example illustrates how `elemAt` is incompatible with the attrpath selector syntax that is the primary means of data exploration.
+Many other programming languages allow syntaxes such as
+`foo.bar.qux[0].moreStuff` (C, many others)
+or `foo.bar.qux(0).moreStuff` (Octave, Scala)
+or `foo.bar.qux.[0].moreStuff` (F#, OCaml)
+or `foo.bar.qux.0.moreStuff` (Coco, LiveScript, the `--attr` option of `nix-instantiate` and other `nix` commands).
+Of these, only `.[]` does not conflict with existing syntax (see [Alternatives] for more details).
+
+# Detailed design
+[design]: #detailed-design
+
+The `attrpath` grammar nonterminal is currently defined as
+
+```
+attrpath
+  : attrpath '.' attr
+  | attrpath '.' string_attr
+  | attr
+  | string_attr
+  ;
+```
+
+where `attr` is a simple identifier and `string_attr` is either a string literal or a bare `${}` interpolation.
+This nonterminal is used in three contexts in the grammar:
+* In a selector (e.g. `expr.attrpath` or `expr.attrpath or default`)
+* As the right-hand side of the `?` operator (e.g. `expr ? attrpath`)
+* As the left-hand side of a binding (e.g. `let attrpath = expr; in body` or `{ attrpath = expr; }`)
+
+This proposal adds two productions to `attrpath`:
+
+```
+  | attrpath '.' '[' INT ']'
+  | '[' INT ']'
+```
+
+(Supporting non-literal expressions is scoped out of this proposal; see [Future work][future].)
+
+In a selector or `?` operator, these new productions have the following semantics:
+* `expr.prefix.[n].suffix` is the result of evaluating `(builtins.elemAt (expr.prefix) n).suffix`
+* `expr ? prefix.[n].suffix` is true if and only if all of the following are true:
+  * `expr ? prefix` (or `prefix` is empty)
+  * `expr.prefix` evaluates to a list with length at least `n + 1`
+  * `expr.prefix.[n] ? suffix` (or `suffix` is empty)
+* `expr.prefix.[n].suffix or expr2` is the result of evaluating `if expr ? prefix.[n].suffix then expr.prefix.[n].suffix else expr2`
+
+It is a syntax error if either new production is used in the left-hand side of a binding.
+
+An implementation of this design is available as patches for Nix at <https://gitlab.com/rhendric/nix-list-index-syntax/>; see instructions there for use.
```

Comment:
I don't expect much regression, but have you tested current nixpkgs trunk?

Reply:
I tested one build on the `nixos-unstable` channel as of yesterday, but not literally trunk.